scripts: mirror-checker check all mirrors

Chris-Peterson444 opened the pull request:
An update to the mirror checker script to add a command for checking all of the country mirrors. The `check-all` command takes all known country codes and then calculates the actual mirrors to test based on DNS records. The script will print out the DNS information for each attempted mirror (i.e., `cc.archive.ubuntu.com`), the calculated real mirrors, and a table with the status of each mirror. It will also save the results of each mirror check in the single mirror check mode style (which is now available under the `check` command), which can be specified with the `-o` flag.

By default it will omit reporting information about mirrors which don't exist (i.e., fallback to `archive.ubuntu.com`) but can be instructed to provide it with the `--all` flag.  

Some examples of the output: https://people.canonical.com/~cpete/mirror-checker-example-output/
- Default output: https://people.canonical.com/~cpete/mirror-checker-example-output/check-all-noble.txt
- All country codes: https://people.canonical.com/~cpete/mirror-checker-example-output/check-all-noble-with-all.txt
- check-all results for us.archive.ubuntu.com: https://people.canonical.com/~cpete/mirror-checker/mirror_checker_output/us.txt


We could clean this up a little more at some point, filtering for real mirrors using DNS information got a little messy. I also expect some more iteration on this if we want to set it up for regular monitoring somewhere.

dbungert commented on scripts/mirror-checker.py:
```diff
@@ -1,77 +1,500 @@
+def get_cname(domain: str) -> str:
+    """Get CNAME records for domain."""
+    try:
+        answers = dns.resolver.Resolver().resolve(domain, "CNAME")
```
suggestion: https://serverfault.com/a/574077 is claiming that it's technically possible but out of spec.  what about:
```suggestion
        [answers] = dns.resolver.Resolver().resolve(domain, "CNAME")
```
which means a return from `resolve()` with more than 1 element will `raise`.

Chris-Peterson444 replied:
Yeah I like this better. Then we could also report information about poorly configured mirrors too.

dbungert commented on scripts/mirror-checker.py:
```diff
@@ -1,77 +1,500 @@
+
+    table = PrettyTable(header)
+    table.add_row(["primary", *mirror_status.pop("primary")])
+    for cc, status in sorted(mirror_status.items()):
```
suggestion: it's great news that all the mirrors look healthy right now, but we expect later that won't be the case.  What do you think about sorting on Status, with `!ok` at the bottom?  It would be nice to have failing mirrors on the bottom, which means we can just check the last lines of the report for failing mirrors and if that isn't the case, we don't need to read the log much further.  WDYT?

Chris-Peterson444 replied:
Yeah that sounds much better to me.

dbungert reviewed (approved):
Thank you for putting this together, this is a great tool to help identify problem archive mirrors.  I have 2 non-blocking suggestions.

Chris-Peterson444 commented:
Thanks for the feedback. I think both suggestions were great and went ahead and implemented them. I'm trying something new and added you as co-author for the commits that include your suggested changes.
